File: 2019/day04.cpp
#include <iostream>
#include <algorithm>
#include <array>

int tests[6] = { 111111, 223450, 123789, 112233, 123444, 111122 };

//ranges = (254032,789860+1)

int valid( int pw, int part )
{
    int dig[6];
    for( int i = 0; i < 6; i++ )
    {
        dig[5-i] = pw % 10;
        pw /= 10;
    }

    std::array<int,10> cnts = { 0 };

    for( int i = 0; i < 6; i++ )
    {
        if( i < 5 && dig[i] > dig[i+1] )
            return 0;
        cnts[dig[i]]++;
    }

    return std::any_of( cnts.begin(), cnts.end(), 
        part == 1 
            ? [](int i){return i>=2;}
            : [](int i){return i==2;}
    );
}


int main()
{
    for( auto t : tests )
    {
        std::cout << t << " P1: " << valid(t,1) << " P2: " << valid(t,2) << "\n";
    }
    
    int cnt1 = 0;
    int cnt2 = 0;
    for( int t = 254032; t <= 789860; t++ )
    {
        cnt1 += valid( t, 1 );
        cnt2 += valid( t, 2 );
    }

    std::cout << "Part 1 : " << cnt1 << "\n"
              << "Part 2 : " << cnt2 << "\n";
}

File: 2019/day04.py
ranges = (254032,789860+1)

def valid(pw,part):
    dig = list(int(k) for k in str(pw))
    cnts = [0]*10
    for c in range(6):
        if c < 5 and dig[c] > dig[c+1]:
            return False
        cnts[dig[c]] += 1

# part 1
#    print( cnts )
    if part == 1:
        return any(k >= 2 for k in cnts )
# part 2
    else:
        return 2 in cnts

tests = ( 111111, 223450, 123789, 112233, 123444, 111122 )

for part in (1,2):
    print( "Part ", part )
    for t in tests:
        print( t,  valid( t, part ) )
     
    cnt = 0
    for r in range( *ranges ):
        if valid(r, part):
            cnt += 1
        
    print( cnt )
